### PR TITLE
Add support for self-hosted runner in CI

### LINKED ISSUE #729: Run CI on self-hosted runner
**Describe the issue**

Use LAL infrastructure instead of GHA runner

**Expected behaviour**

Run science pipeline on CI,

**Actual behaviour**

Science pipeline cannot run on GHA action because it requires to much resources (storage, memory)

### PR DESCRIPTION
**Pull request by fjammes**
Linked to issue(s): Closes #729 

## What changes were proposed in this pull request?

Important features implemented:
- Wait for data on spark storage instead of crashing in raw2science and distribute drivers
- Wait for topics to appear in order to validate integration tests have passed in CI
- Add support for kafka AuthN

And also:
- Override POD_NAMESPACE on CI node
- Run CI with the science image
- Add support for reusable workflow
- Add support for secrets
- Use self-hosted runner internal registry
- Add integration test with science algorithms
- Test insecure registry setup in kind
- Add conditional import of science modules
- Fail CI if the number of spark pods is not correct
- Remove column 'lc_features_g' from noscience pipeline
- Disable apply_user_defined_filter() in noscience pipeline
- Install jdk in CI
- Increase fink startup control delay
- Add default value for fink-broker '--exit-after' option
- Fail docker build if not able to curl a jar file
- Set kafka volumes size to 1Gi
- Use pvc for minio
- Add user aliases
- Disable AuthZ in kafka
- Enable GHA ci for noscience pipeline

## How was this patch tested?

The "noscience" pipeline works fine on GHA, but there is still an issue with the docker registry on self-hosted runner. However "noscience" and "science" integration tests have passed successfully on the self-hosted runner.

### CODE REVIEW

File: fink_broker/sparkUtils.py [hunk 1]
```diff
@@ -326,6 +333,27 @@ def connect_to_raw_database(
 
     return df
 
+def path_exist(path: str) -> bool:
+    """Check if a path exists on Spark FS
```
Comment (JulienPeloton) [inline fink_broker/sparkUtils.py:337]: nit: ... exists on HDFS

File: fink_broker/sparkUtils.py [hunk 2]
```diff
@@ -326,6 +333,27 @@ def connect_to_raw_database(
...
+    path : str
+        Path to check
+
+    Returns
+    -------
+    bool
+        True if the path exists, False otherwise
+    """
+    sc = SparkContext._active_spark_context
+    try:
```
Comment (JulienPeloton) [inline fink_broker/sparkUtils.py:350]: I suggest to get rid of the ``try/except`, and use 

```python
def path_exist(path: str) -> bool:
        """ ...
        """
        spark = SparkSession \
                .builder \
                .getOrCreate()

        jvm = spark._jvm
        jsc = spark._jsc 

        fs = jvm.org.apache.hadoop.fs.FileSystem.get(jsc.hadoopConfiguration()) 
        if fs.exists(jvm.org.apache.hadoop.fs.Path(path)): 
                return True
        return False
```

**Review by JulienPeloton — COMMENTED** [review]
Thanks @fjammes ! At first glance, it looks good to me. A lot of new interesting stuff :-) 
I left some comments.

File: .github/workflows/itest-gha.yml [hunk 3]
```diff
@@ -0,0 +1,164 @@
+# Qserv operator CI workflow
```
[the remark this answers is not in the document]
Comment (fjammes) [reply .github/workflows/itest-gha.yml:1]: :+1:

File: TODO.org [hunk 4]
```diff
@@ -1,11 +1,28 @@
...
+** TODO DELAYED BECAUSE IT NOT BLOCKS BUT WARN create topic in distribute before sending alerts in order to avoid error below: https://fink-broker.slack.com/archives/D03KJ390F17/p1692008729660549
+Du coup ça fonctionne avec un compte utilisateur, par contre j'ai pas activé les autorisations dans kafka car le fink-alert-simulator aurait pu plus écrire dans le topic sans authentification.
+12 h 28
+J'ai maintenant ce message d'erreur:
+23/08/14 10:26:52 WARN NetworkClient: [Producer clientId=producer-1] Error while fetching metadata with correlation id 29 : {fink_simbad_grav_candidates_ztf=LEADER_NOT_AVAILABLE}
+12 h 32
+En fait c'est du au fait que le topic existe pas, ça fonctionne si on relance lae job distribute...
+12 h 33
+Tu crois qu'on pourrais pré-créer les topic pour éviter ce problème
+@JulienPeloton
```
[the remark this answers is not in the document]
Comment (fjammes) [reply TODO.org:12]: This is not blocking, it only prints a warning. So I propose we cancel this proposal.

File: fink_broker/sparkUtils.py [hunk 5]
```diff
@@ -309,6 +312,10 @@ def connect_to_raw_database(
         .builder \
         .getOrCreate()
 
+    while not path_exist(basepath):
+        print("Waiting for data to arrive in {}".format(basepath))
+        time.sleep(5)
```
[the remark this answers is not in the document]
Comment (fjammes) [reply fink_broker/sparkUtils.py:317]: I have added a wait time which increase of 20% at each iteration, start from 5 seconds (useful for e2e tests and CI) and stop increasing when it has reached 60 sec.

File: fink_broker/sparkUtils.py (hunk 1 above)
Comment (fjammes) [reply fink_broker/sparkUtils.py:337]: Not only HDFS, it might also be S3 (Minio in CI for example)

File: fink_broker/sparkUtils.py (hunk 2 above)
Comment (fjammes) [reply fink_broker/sparkUtils.py]: Hi @JulienPeloton , this code does not work out of the box, I have the following error message:
```
23/09/05 12:44:36 INFO SharedState: Warehouse path is 'file:/home/fink/spark-warehouse'.
Traceback (most recent call last):
  File "/home/fink/fink-broker/bin/distribute.py", line 154, in <module>
    main()
  File "/home/fink/fink-broker/bin/distribute.py", line 71, in main
    df = connect_to_raw_database(
  File "/home/fink/fink-broker/fink_broker/sparkUtils.py", line 315, in connect_to_raw_database
    while not path_exist(basepath):
  File "/home/fink/fink-broker/fink_broker/sparkUtils.py", line 374, in path_exist
    if fs.exists(jvm.org.apache.hadoop.fs.Path(path)):
  File "/opt/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py", line 1321, in __call__
  File "/opt/spark/python/lib/pyspark.zip/pyspark/sql/utils.py", line 117, in deco
pyspark.sql.utils.IllegalArgumentException: Wrong FS: s3a://fink-broker-online/science/year=2020/month=01/day=01, expected: file:///
23/09/05 12:44:39 WARN ExecutorPodsWatchSnapshotSource: Kubernetes client has been closed.
```
Are you sure this code is compliant with S3 storage?

Comment (JulienPeloton) [reply fink_broker/sparkUtils.py]: ah good point! For both HDFS & S3, one should provide URI. This should work:

```python
def path_exist(path: str) -> bool:
        """ ...
        """
        spark = SparkSession \
                .builder \
                .getOrCreate()

        jvm = spark._jvm
        jsc = spark._jsc 

        conf = sc._jsc.hadoopConfiguration()
        uri = jvm.java.net.URI(path)

        fs = jvm.org.apache.hadoop.fs.FileSystem.get(uri, conf)
        if fs.exists(jvm.org.apache.hadoop.fs.Path(path)): 
                return True
        return False
```

Comment (fjammes) [reply fink_broker/sparkUtils.py]: I suppose I need to replace `conf = sc._jsc.hadoopConfiguration()` with  `conf = jsc.hadoopConfiguration()` ?

Comment (fjammes) [reply fink_broker/sparkUtils.py]: This seems to work on my dev server, let's see if it works in CI: https://github.com/astrolabsoftware/fink-broker/actions/runs/6086058413/job/16511542608

Comment (JulienPeloton) [reply fink_broker/sparkUtils.py]: Yes, `jsc` or `sc` should work equivalently. The former is just the Java object directly.

**Pull request merged**